narwhals/series.py

```diff
@@ -157,6 +158,71 @@ def from_numpy(
         )
         raise ValueError(msg)
 
+    @classmethod
+    def from_iterable(
+        cls,
+        name: str,
+        values: Iterable[Any],
+        dtype: IntoDType | None = None,
+        *,
+        backend: ModuleType | Implementation | str,
+    ) -> Series[Any]:
+        """Construct a Series from an iterable.
+
+        Arguments:
+            name: Name of resulting Series.
+            values: One-dimensional data represented as an iterable.
+            dtype: (Narwhals) dtype. If not provided, the native library
+                may auto-infer it from `values`.
+            backend: specifies which eager backend instantiate to.
+
+                `backend` can be specified in various ways
+
+                - As `Implementation.<BACKEND>` with `BACKEND` being `PANDAS`, `PYARROW`,
+                    `POLARS`, `MODIN` or `CUDF`.
+                - As a string: `"pandas"`, `"pyarrow"`, `"polars"`, `"modin"` or `"cudf"`.
+                - Directly as a module `pandas`, `pyarrow`, `polars`, `modin` or `cudf`.
+
+        Returns:
+            A new Series
+
+        Examples:
+            >>> import pandas as pd
+            >>> import narwhals as nw
+            >>>
+            >>> values = [4, 1, 3, 2]
+            >>> nw.Series.from_iterable("a", values, dtype=nw.UInt32, backend="pandas")
+            ┌──────────────────────┐
+            |   Narwhals Series    |
+            |----------------------|
+            |0    4                |
+            |1    1                |
+            |2    3                |
+            |3    2                |
+            |Name: a, dtype: uint32|
+            └──────────────────────┘
+        """
+        if is_numpy_array(values):
+            return cls.from_numpy(name, values, dtype, backend=backend)
+        if dtype:
+            _validate_into_dtype(dtype)
+        if not isinstance(values, Iterable):
+            msg = f"Expected values to be an iterable, got: {qualified_type_name(values)!r}."
+            raise TypeError(msg)
+        implementation = Implementation.from_backend(backend)
+        if is_eager_allowed(implementation):
+            ns = cls._version.namespace.from_backend(implementation).compliant
+            compliant = ns._series.from_iterable(
+                values, context=ns, name=name, dtype=dtype
+            )
+            return cls(compliant, level="full")
+        msg = (
```

Comment:
Should we add a similar tentative as in `new_series` with `Implementation.UNKNOWN`?

```suggestion
        if implementation is Implementation.UNKNOWN:  # pragma: no cover
            _native_namespace = implementation.to_native_namespace()
            try:
                native_series: NativeSeries = _native_namespace.from_iterable(
                    name=name, values=values, dtype=dtype
                )
                return from_native(native_series, series_only=True).alias(name)
            except AttributeError as e:
                msg = "Unknown namespace is expected to implement `from_iterable` constructor."
                raise AttributeError(msg) from e
        msg = (
```

Reply:
I think I wrote about this in one of the other `<class>.from_<thing>` PRs

Due to (#2786) - I don't think it is helpful to introduce the unreachable `Implementation.UNKNOWN` paths in this version

I would strongly prefer we have that functionality *working* from the start here and fully covered by tests

```py
import narwhals as nw

name = "a"
values = [4, 1, 2, 3]
backend = nw.Implementation.UNKNOWN
>>> nw.new_series(name, values, backend=backend)
```

```py
File c:\Users\danie\Documents\GitHub\narwhals\narwhals\functions.py:227, in _new_series_impl(name, values, dtype, backend)
    225     return series.to_narwhals()
    226 if implementation is Implementation.UNKNOWN:  # pragma: no cover
--> [227](file:///C:/Users/danie/Documents/GitHub/narwhals/narwhals/functions.py:227)     _native_namespace = implementation.to_native_namespace()
    228     try:
    229         native_series: NativeSeries = _native_namespace.new_series(
    230             name, values, dtype
    231         )

File c:\Users\danie\Documents\GitHub\narwhals\narwhals\_utils.py:378, in Implementation.to_native_namespace(self)
    376 if self is Implementation.UNKNOWN:
    377     msg = "Cannot return native namespace from UNKNOWN Implementation"
--> [378](file:///C:/Users/danie/Documents/GitHub/narwhals/narwhals/_utils.py:378)     raise AssertionError(msg)
    380 self._backend_version()
    381 module_name = _IMPLEMENTATION_TO_MODULE_NAME.get(self, self.value)

AssertionError: Cannot return native namespace from UNKNOWN Implementation
```

I hope that makes sense, I'm just deferring the fix to (#2786) and only re-implementing the working parts of what these constructors supersede 🙂

Reply:
Fair enough! Thank for explaining in details what's happening